src/Build.UnitTests/Graph/IsolateProjects_Tests.cs

```diff
@@ -59,51 +90,248 @@ public void CacheEnforcementShouldAcceptSelfReferences(BuildResultCode expectedB
+        public void CacheAndTaskEnforcementShouldAcceptCallTarget()
         {
-            AssertBuild(new []{"BuildReference"},
+            AssertBuild(new []{"CallTarget"},
+                (result, logger) =>
+                {
+                    result.OverallResult.ShouldBe(BuildResultCode.Success);
+
+                    logger.Errors.ShouldBeEmpty();
+                });
+        }
+
+        [Fact(Skip = "https://github.com/Microsoft/msbuild/issues/3876")]
+        public void CacheEnforcementShouldFailWhenReferenceWasNotPreviouslyBuiltAndOnContinueOnError()
+        {
+            CacheEnforcementShouldFailWhenReferenceWasNotPreviouslyBuilt2(true);
+        }
+
+        [Fact]
+        public void CacheEnforcementShouldFailWhenReferenceWasNotPreviouslyBuiltWithoutContinueOnError()
+        {
+            CacheEnforcementShouldFailWhenReferenceWasNotPreviouslyBuilt2(false);
+        }
+
+        private void CacheEnforcementShouldFailWhenReferenceWasNotPreviouslyBuilt2(bool addContinueOnError)
+        {
+            AssertBuild(
+                new[] {"BuildDeclaredReference"},
                 (result, logger) =>
                 {
                     result.OverallResult.ShouldBe(BuildResultCode.Failure);
 
                     logger.ErrorCount.ShouldBe(1);
 
-                    logger.Errors.First().Message.ShouldStartWith("MSB4252:");
-                });
+                    logger.Errors.First()
+                        .Message.ShouldStartWith("MSB4252:");
+                },
+                addContinueOnError: addContinueOnError);
         }
 
         [Fact]
         public void CacheEnforcementShouldAcceptPreviouslyBuiltReferences()
         {
-            AssertBuild(new []{"BuildReference"},
+            AssertBuild(new []{"BuildDeclaredReference"},
                 (result, logger) =>
                 {
                     result.OverallResult.ShouldBe(BuildResultCode.Success);
 
                     logger.Errors.ShouldBeEmpty();
                 },
-                buildReference: true);
+                buildDeclaredReference: true);
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void TaskEnforcementShouldFailOnUndeclaredReference(bool addContinueOnError)
+        {
+            AssertBuild(new[] { "BuildUndeclaredReference" },
+                (result, logger) =>
+                {
+                    result.OverallResult.ShouldBe(BuildResultCode.Failure);
+
+                    logger.ErrorCount.ShouldBe(1);
+
+                    logger.Errors.First().Message.ShouldStartWith("MSB4254:");
+                },
+                addContinueOnError: addContinueOnError);
         }
 
-        private void AssertBuild(string[] targets, Action<BuildResult, MockLogger> assert, bool buildReference = false)
+        [Fact]
+        public void TaskEnforcementShouldNotAcceptPreviouslyBuiltReferences()
```

Comment:
I'm not sure I understand this test name. Why shouldn't it accept this reference? Because it's previously built but also undeclared for this project?

Reply:
Will rename to `TaskEnforcementShouldFailOnPreviouslyBuiltButUndeclaredReferences`